src/validation_layer.h

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_VALIDATION_LAYER_H
+#define BITCOIN_VALIDATION_LAYER_H
+
+#include <future>
+
+#include <chainparams.h>
+#include <core/consumerthread.h>
+#include <core/producerconsumerqueue.h>
+#include <util.h>
+
+class ValidationLayer;
+extern std::unique_ptr<ValidationLayer> g_validation_layer;
+
+/**
+ * Encapsulates a request to validate an object (currently only a block)
+ * Submitted to ValidationLayer for asynchronous validation
+ *
+ * @see ValidationLayer
+ */
+template <typename RESPONSE>
+class ValidationRequest : public WorkItem<WorkerMode::BLOCKING>
+{
+    friend ValidationLayer;
+
+private:
+    //! Guts of the validation
+    virtual void operator()() = 0;
+
+    //! Returns a string identifier (for logging)
+    virtual std::string GetId() const = 0;
+
+protected:
+    //! Promise that will deliver the validation result to the caller who generated this request
+    std::promise<RESPONSE> m_promise;
+};
+
+/**
+ * Holds the results of asynchronous block validation
+ */
+struct BlockValidationResponse {
+    //! Is this the first time this block has been validated
+    const bool is_new;
+
+    //! Did initial validation pass (a block can still pass initial validation but then later fail to connect to an active chain)
+    const bool block_valid;
```

Comment:
Based on the call site, should this be `block_accepted`? "valid" is a little ambiguous if you take into account all the different states in `chain.h:BlockStatus`.